solidity/src/CadenceRandomConsumer.sol

```diff
@@ -56,6 +101,85 @@ abstract contract CadenceRandomConsumer is CadenceArchWrapper {
     }
 
+    /**
+     * ----- REVEAL STEP -----
+     */
+
+    /**
+     * @dev This method fulfills a random request and returns a random number as a uint64.
+     *
+     * @param requestId The ID of the randomness request to fulfill.
+     * @return randomResult The random number.
+     */
+    function _fulfillRandomRequest(uint256 requestId) internal returns (uint64) {
+        bytes32 randomValue = _fulfillRandomness(requestId);
+        uint64 randomResult = uint64(uint256(keccak256(abi.encodePacked(randomValue, requestId))));
+
+        emit RandomnessFulfilled(requestId, randomResult);
+
+        return randomResult;
+    }
+
+    /**
+     * @dev This method fulfills a random request and safely returns an unbiased random number in the range [min, max].
+     *
+     * @param requestId The ID of the randomness request to fulfill.
+     * @param min The minimum value of the range (inclusive).
+     * @param max The maximum value of the range (inclusive).
+     * @return randomResult The random number in the range [min, max].
+     */
+    function _fulfillRandomInRange(uint256 requestId, uint64 min, uint64 max) internal returns (uint64) {
+        // Ensure that the request is fulfilled at a Flow block height greater than the one at which the request was made
+        // Get the random source for the Flow block at which the request was made
+        bytes32 randomSource = _fulfillRandomness(requestId);
+        // Pack the randomResult into a uint256, hashing with the requestId to vary results across shared block heights.
+        uint256 randomValue = uint256(keccak256(abi.encodePacked(randomSource, requestId)));
+
+        uint64 randomResult = _getNumberInRange(randomValue, min, max);
+
+        emit RandomnessFulfilled(requestId, randomResult);
+
+        return randomResult;
+    }
+
+    ////////////////////
+    // PRIVATE FUNCTIONS
+    ////////////////////
+
+    /**
+     * @dev This method returns a number in the range [min, max] from the given value.
+     * NOTE: You may be tempted to simply use value % (max - min + 1) to get a number in the range [min, max]. However,
+     * this method is not secure is susceptible to the modulo bias. This method provides an unbiased alternative for
+     * secure use of randomness.
+     *
+     * @param min The minimum value of the range (inclusive).
+     * @param max The maximum value of the range (inclusive).
+     * @return random The random number in the range [min, max].
+     */
+    function _getNumberInRange(uint256 value, uint64 min, uint64 max) private pure returns (uint64) {
+        require(max > min, "Max must be greater than min");
+
+        uint64 range = max - min + 1;
+        uint64 bitsRequired = _mostSignificantBit(range - 1); // Number of bits needed to cover the range
+        uint256 mask = (1 << bitsRequired) - 1; // Create a bitmask to extract relevant bits
+        uint64 candidate = 0; // Initialize candidate
+
+        while (true) {
```

Comment:
There seems to be an issue because `value` is fixed size (256 bits), so we may run out of randoms before we find a real random `candidate` less than `range`. 
The loop would eventually return because `value` will become zero at some point (it's repeatedly being shifted to the right), and a zero `candidate` is always less than `range`  but that's not the behaviour we want.

This is the perfect example that requires a PRG: we start from a random source of 32 bytes, but we want an arbitrary number of bytes out of that source. So instead of hashing, we use a PRG that generates as many pseudo-random bytes as we want

Reply:
> The loop would eventually return because value will become zero at some point (it's repeatedly being shifted to the right), and a zero candidate is always less than range but that's not the behaviour we want.

I attempt to cover this case with the line

```solidity
            require(value > 0, "Random number source exhausted");
```

but wondered myself if that's against the expectation that `min` is inclusive

Reply:
I also missed the `require`, but what will happen in that case? the function errors and there won't be a random?

Reply:
The contract gets 128 bits of entropy as an input, which is a safe amount to generate lots of random data. If the computed hash didn't lead to finding the desired output the contract should continue looking for one, using the same source. That's why I think the PRG use is still relevant (unless I missed another thing in the code)